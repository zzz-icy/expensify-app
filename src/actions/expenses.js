// Actions for expenses
// import uuid from 'uuid';
import database from '../firebase/firebase';
// action generators
// ADD-EXPENSE

// export const addExpense = (
//     { description = '', note = '', amount = 0, createdAt = 0 } = {}
// ) => (
//         {
//             type: 'ADD_EXPENSE',
//             expense: {
//                 id: uuid(),
//                 description,
//                 note,
//                 amount,
//                 createdAt,
//             }
//         });

// we do not need the id generated by uuid(), will use the id generated by firebase, so rewrite the addExpense

export const addExpense = (expense) => (
    {
        type: 'ADD_EXPENSE',
        expense
    });


export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };
        // add return so that we can chain .then() in the test
        return database.ref('expenses').push(expense).then((ref) => { //.then() gets called with the ref, so we have access to the id 
            dispatch(addExpense(
                {
                    id: ref.key,
                    ...expense
                }
            ));
        });
    }
};


// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

export const startRemoveExpense = ({ id }) => {
    return (dispatch) => {
        return database.ref(`expenses/${id}`).remove().then(() => {
            // once it is removed from database, dispatch action
            dispatch(removeExpense({ id }));
        });
    }
}
// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

// 1. fetch all expenses data
// 2. Parse that data into an array, cause data from firebase are nested objects with the id as the key for each expense
// 3. dispatch SET_EXPENSES

export const startSetExpenses = () => {
    return (dispatch) => {

        // add return so that we can chain .then() in the app.js
        return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            });
            dispatch(setExpenses(expenses));
        });
    }
};