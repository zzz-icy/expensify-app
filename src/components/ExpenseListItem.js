import React from 'react';
import moment from 'moment';
import numeral from 'numeral';
import { Link } from 'react-router-dom';

const ExpenseListItem = (props) => {
    // const expenses = props.expenses;
    // console.log(props);
    return (
        <div>
            <Link to={`/edit/${props.data.id}`}><h3> {props.data.description} </h3></Link>
            <p>
                {numeral(props.data.amount / 100).format('$0,0.00')}
                -
                {moment(props.data.createdAt).format('MMMM Do, YYYY')}
            </p>

        </div>
    );
};

// deconstructure

//<ExpenseListItem data={item} {...item} />
// if the data is passed in like {...item}, in ExpenseListItem will can use the destructure version

// const ExpenseListItem = ({ description, amount, createdAt }) => {
//     // const expenses = props.expenses;
//     // console.log(props);
//     return (
//         <div>
//             <h3> {description} </h3>
//             <p> {amount} - {createdAt}</p>
//         </div>
//     );
// };

// we do not need the states in store, just want the dispatch, no arguments passed in 
export default ExpenseListItem;