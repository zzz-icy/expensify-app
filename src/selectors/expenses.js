// get visible expenses


const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {

    return expenses.filter(
        (item) => {
            const startDateMatch = typeof startDate !== 'number' || item.createdAt >= startDate;
            const endDateMatch = typeof endDate !== 'number' || item.createdAt <= endDate;
            const textMatch = item.description.toLowerCase().includes(text.toLowerCase());

            // figure out if expenses.description as the text variable string inside of it
            // includes, The includes() method determines whether an array includes a certain element, returning true or false as appropriate.
            // convert both string to lower case

            // sort, in place and retur nthe array
            // sort(compareFunction(a,b)) 
            // return 1, b to lower index
            // return 0, unchanged
            // return -1, a to lower index
            // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort
            return startDateMatch && endDateMatch && textMatch;

        }
    ).sort(
        (a, b) => {
            if (sortBy === 'date') {
                return (a.createdAt < b.createdAt) ? 1 : -1;// latest date first
            }
            if (sortBy === 'amount') {
                return (a.amount < b.amount) ? 1 : -1; // higher amount first
            }
        }
    );
};

export default getVisibleExpenses;