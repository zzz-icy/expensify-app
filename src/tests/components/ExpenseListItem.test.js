// react-test-renderer
import React from 'react';
// import ReactShallowRenderer from 'react-test-renderer/shallow'; // ES6
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem'; // we just want the unconnected version
import "../setupTests.js";
import expenses from '../fixtures/expenses';
// Convert Enzyme wrappers to a format compatible with Jest snapshot testing.
test('should render expense list item properly', () => {
    const wrapper = shallow(<ExpenseListItem data={expenses[0]} key={expenses[0].id} />);
    expect(wrapper).toMatchSnapshot();

});