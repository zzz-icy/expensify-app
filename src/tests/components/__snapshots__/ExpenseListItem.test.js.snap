// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list item properly 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
       
      bill
       
    </h3>
  </Link>
  <p>
    $5.00
    -
    December 31st, 1969
  </p>
</div>
`;
